net: ot: OpenThread HOST to RCP communication via UART

lmajewski opened the pull request:
This patch set is in some sense the follow up for #75986 - however it is much simpler as it uses UART (serial port) to communicate via SPINEL protocol with RCP device.

This patch set is agnostic to architecture on which it is running - i.e. it uses Zephyr's API for serial port communication (NXP has for example `SerialManager` program in their HAL).

The OpenThread's RX callback is called from a system based workqueue after the serial IRQ is served.

Used HW:
-  nRF21540-DK (Radio Co-Processor) `west build -p always -b nrf21540dk/nrf52840 -S logging nrf/samples/openthread/coprocessor` from Nordic SDK
- mimxrt1020-EVK, which is the HOST

Another nRF21540-DK with (ot-cli) is used as RCP for PC controlled setup to check if OpenThread's communication works as expected.

The `echo_client` program has been used with proper DTS and configuration overlays.
`west build -p always -b mimxrt1020_evk samples/net/sockets/echo_client -- -DCONF_FILE="prj.conf overlay-ot-rcp-host-uart.conf"`
Following code can be used to check if the communication works correctly:

**Lider/Commissioner (ot-cli):**(nRF21540-DK)

`dataset init new`
`dataset`
`dataset commit active`
`ifconfig up`
`thread start`

`commissioner start`
`commissioner joiner add * J01NME`
**Child:**(imxrt1020-evk)
`ot factoryreset`
`ot ifconfig up`
`ot joiner start J01NME`
`ot thread start`
`net ping -c 10000 fdc2:f018:d073:9b12:0:ff:fe00:fc00 `

Commit:
net: samples: ot: Add configuration file for echo_client OT RCP HOST (UART)

This config allows setting up the echo_client program with
RCP HOST UART (via SPINEL protocol) communication.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
net: samples: Add mimxrt1020_evk overlay for OT's RCP Host support (UART)

This overlay configures the imxrt1020 EVK board to use its lpuart2 serial
port to communicate with other board (in this case nRF21540-DK), which
is the RCP for OpenThread protocol.

The nRF Zephyr RCP is the one from Nordic (v2.7.0):
west build -p always -b nrf21540dk/nrf52840 -S logging
		nrf/samples/openthread/coprocessor

The same UART (on nRF) can be used for testing/debugging purposes with
using the ot-cli's SPINEL communication program:
sudo ./build/posix/src/posix/ot-cli 'spinel+hdlc+uart:///dev/ttyACM1?u
art-baudrate=1000000' --verbose

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: ot: Add HDLC HOST support with UART communication to RCP

This patch adds support for HOST OpenThread communication to the RCP
co-processor via UART using SPINEL protocol.

The aim is to use OpenThread's RCP (Radio Co-Processor) with HOST device
(for example imxRT1020). Such configuration is the same as one used
with PC program (ot-cli) and RCP.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

rlubos commented on samples/net/sockets/echo_client/overlay-ot-rcp-host-uart.conf:
```diff
@@ -0,0 +1,49 @@
+CONFIG_REQUIRES_FULL_LIBC=y
```
The instructions from commit message 2487f352b9e3ae2a9d546378274986e58581a681 or 5421b134943a19ba97939ff56b05d92561efe74e could be included somewhere in the readme file, I doubt anyone would look for such info in the commit history.

rlubos commented on drivers/hdlc_rcp_if/hdlc_rcp_if_uart.c:
```diff
@@ -0,0 +1,216 @@
+	if (len > 0) {
+		rd_len = uart_fifo_read(dev, data, len);
+
+		int err = ring_buf_put_finish(
+			ot_uart.rx_ringbuf, rd_len);
```
This should fit into 80 chars/line limit, I guess no need to break the line.
I'm also a bit puzzled on why compliance doesn't "complain" it typically requires an empty line after variable definition.

rlubos reviewed (dismissed):
Some nits, but overall looks good. Thanks for the contribution!

Commit:
net: Extend echo_client README.rst file to setup OTBR based echo_server

This description provides information about setting up RCP (nRF21540-DK)
and OT HOST (mimxrt1020_evk) to communicate with HOST PC's echo_server
via OpenThread Border Router (OTBR).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
Changes for v2:
- Use Ring Buffer for data to be transmitted to OT network (the OpenThread allows max packet size to be 1300 bytes)
- Move the testing documentation from commit messages to README.rst
- Use OTBR docker container to allow communication between on-HOST `echo_server` and on-target (RCP+imxRT1020) `echo_client`
- Use workqueue to run OpenThread callback.
